src/index.ts

```diff
@@ -10,67 +10,97 @@ import {
 import {BeaconEventType} from "./types";
 import {verifyArgs} from "./utils";
 import {Epoch} from "@chainsafe/lodestar-types";
-import {CID_FILE_PATH} from "./constants";
+import {CID_FILE_PATH, WS_STATE_DATA_FILE_PATH, WAITING_MSG} from "./constants";
+import {toHexString} from "@chainsafe/ssz";
+
+let alreadyFetchingState = false;
+
+async function getAndUploadState(wsEpoch: Epoch): Promise<void> {
+  console.log(`Getting state for weak subjectivity epoch ${wsEpoch}...`);
+  let state;
+  try {
+    state = await getBeaconStateBuffer(config, wsEpoch);
+  } catch (error) {
+    throw new Error(
+      `State retrieval/storage error for weak subjectivity epoch ${wsEpoch}.  Retrying.  Reason: ${error.message}`
+    );
+  }
+  if (!state) throw new Error(`State at weak subjectivity epoch ${wsEpoch} not found`);
+  console.log(`Found state for weak subjectivity epoch ${wsEpoch}`);
+
+  const cid = await uploadState(state, wsEpoch);
+
+  const stateRoot = toHexString(
+    config.types.phase0.BeaconState.hashTreeRoot(config.types.phase0.BeaconState.deserialize(state))
+  );
+  const stateData = {stateRoot, ipfsPath: "/ipfs/" + cid + "/state.ssz"};
+
+  // store stateData locally
+  fs.writeFileSync(WS_STATE_DATA_FILE_PATH, JSON.stringify(stateData, null, 2), "utf-8");
+  
+  // store IPFS hash (CID) in local file
+  fs.writeFileSync(CID_FILE_PATH, cid, "utf-8");
+
+  console.log(WAITING_MSG);
+  alreadyFetchingState = false;
```

Comment:
Can you ensure this lock will be released on any possible codepath including all errors and throws?